sqldb/invoices.go

```diff
@@ -0,0 +1,1554 @@
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		// Add context to unique constraint errors.
+		var uniqueConstraintErr *ErrSQLUniqueConstraintViolation
+		if errors.As(err, &uniqueConstraintErr) {
+			return 0, invpkg.ErrDuplicateInvoice
+		}
+
+		return 0, fmt.Errorf("unable to add invoice(%v): %w",
+			paymentHash, err)
+	}
+
+	newInvoice.AddIndex = uint64(invoiceID)
+
+	return newInvoice.AddIndex, nil
+}
+
+// fetchInvoiceData fetches the common invoice data and the AMP state for the
+// invoice with the given reference.
+func (i *InvoiceStore) fetchInvoice(ctx context.Context,
+	db InvoiceQueries, ref invpkg.InvoiceRef) (*invpkg.Invoice, error) {
+
+	if ref.PayHash() == nil && ref.PayAddr() == nil && ref.SetID() == nil {
+		return nil, invpkg.ErrInvoiceNotFound
+	}
+
+	var (
+		invoice *invpkg.Invoice
+		params  sqlc.GetInvoiceParams
+	)
+
+	if ref.PayHash() != nil {
+		params.Hash = ref.PayHash()[:]
+	}
+
+	if ref.PayAddr() != nil && *ref.PayAddr() != invpkg.BlankPayAddr {
+		params.PaymentAddr = ref.PayAddr()[:]
+	}
+
+	// Fetch all AMP HTLCs, unless it is explicitly requested to only fetch
+	// the AMP HTLCs for this specific set.
+	if ref.SetID() != nil && ref.Modifier() != invpkg.HtlcSetOnlyModifier {
+		params.SetID = ref.SetID()[:]
+	}
+
+	rows, err := db.GetInvoice(ctx, params)
+	switch {
+	case len(rows) == 0:
+		return nil, invpkg.ErrInvoiceNotFound
+
+	case len(rows) > 1:
+		// In case the reference is ambiguous, meaning it matches more
+		// than	one invoice, we'll return an error.
+		return nil, fmt.Errorf("ambiguous invoice ref: %s",
+			ref.String())
+
+	case err != nil:
+		return nil, fmt.Errorf("unable to fetch invoice: %w", err)
+	}
+
+	var (
+		setID         *[32]byte
+		fetchAmpHtlcs bool
+	)
+
+	// Now that we got the invoice itself, fetch the HTLCs as requested by
+	// the modifier.
+	switch ref.Modifier() {
+	case invpkg.DefaultModifier:
+		// By default we'll fetch all AMP HTLCs.
+		setID = nil
+		fetchAmpHtlcs = true
+
+	case invpkg.HtlcSetOnlyModifier:
+		// In this case we'll fetch all AMP HTLCs for the
+		// specified set id.
+		if ref.SetID() == nil {
+			return nil, fmt.Errorf("set ID is required to use " +
+				"the HTLC set only modifier")
+		}
+
+		setID = ref.SetID()
+		fetchAmpHtlcs = true
+
+	case invpkg.HtlcSetBlankModifier:
+		// No need to fetch any HTLCs.
+		setID = nil
+		fetchAmpHtlcs = false
+
+	default:
+		return nil, fmt.Errorf("unknown invoice ref modifier: %v",
+			ref.Modifier())
+	}
+
+	// Fetch the rest of the invoice data and fill the invoice struct.
+	_, invoice, err = fetchInvoiceData(
+		ctx, db, rows[0], setID, fetchAmpHtlcs,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return invoice, nil
+}
+
+func fetchAmpState(ctx context.Context, db InvoiceQueries, invoiceID int64,
+	setID *[32]byte, fetchHtlcs bool) (invpkg.AMPInvoiceState,
+	invpkg.HTLCSet, error) {
+
+	var paramSetID []byte
+	if setID != nil {
+		paramSetID = setID[:]
+	}
+
+	ampInvoiceRows, err := db.FetchAMPInvoices(
```

Comment:
Should only be a single AMP invoice per AMP invoice IMO. However, there can be many payments to a single AMP invoice, unlike regular MPP invoices.